src/metabase/util/ordered_hierarchy.clj

```diff
@@ -0,0 +1,131 @@
+    (if (contains? processed n)
+      state
+      (let [children   (get g n)
+            init-state (update state :processing conj n)
+            post-state (reduce (partial toposort-visit g)
+                               init-state
+                               ;; Iterate in reverse as that gives us oldest to newest
+                               (reverse children))]
+        (-> post-state
+            (update :processed conj n)
+            (update :processing disj n))))))
+
+(defn- toposort
+  "Use Trajan's DPS method to return an ordered set of nodes where every node is preceded by all its children, and
+  those children are likewise enumerated in the order they were registered as children of that node."
+  [roots graph]
+  (->> roots
+       (reduce (partial toposort-visit graph)
+               {:processing #{}
+                :processed  (ordered-set)})
+       :processed))
+
+(defn sorted-tags
+  "An ordered set of all tags within the hierarchy, topologically sorted from the leaves up, and following the edges
+  according to the order they were added in."
+  [h]
+  (let [roots (reduce disj
+                      (into (ordered-set) (keys (:children h)))
+                      (keys (:parents h)))]
+    (toposort roots (:children h))))
+
+(defn- calculate-derived-fields [h]
+  (let [ts          (sorted-tags h)
+        re-sort     (fn [m sorted] (m/map-vals #(into (ordered-set) (filter % sorted)) m))
+        ancestors   (re-sort (:ancestors h) ts)
+        descendants (re-sort (:descendants h) (reverse ts))]
+    (assoc h
+      :sorted-tags ts
+      :ancestors ancestors
+      :descendants descendants)))
+
+(defn derive
+  "Establishes a parent/child relationship between two keyword tags, similar to [[clojure.core/derive]].
+  Where it differs is that the order in which we derive keys is significant - it determines the order in which we return
+  its parents. Ancestors are returned according to a toposort, and descendants are returned in the opposite order."
+  [h tag parent]
+  (assert (not= tag parent))
+  (assert (keyword tag))
+  (assert (keyword? parent))
+  (assert (::ordered? (meta h)) "This operation requires an ordered hierarchy.")
+
+  (let [tp (:parents h)
+        tc (:children h)]
+    (if (contains? (tp tag) parent)
+      h
+      (-> (assoc h :parents (update tp tag #(or % (ordered-set))))
+          (clojure.core/derive tag parent)
+          (assoc :children (update tc parent #(into (ordered-set tag) %)))
+          calculate-derived-fields
+          (with-meta {::ordered? true})))))
+
+(defn- first-common-tag
+  "Given two ordered sets corresponding to ancestor in a hierarchy, return the first ancestor of tag-a which is also an
+  ancestor of tag-b.
+  Returns nil if there is no such intersection."
+  [ancestors-a ancestors-b]
+  ;; I suspect this is not commutative, so this ordering is important - we care more about being close to tag-a.
+  (some ancestors-b ancestors-a))
+
+(defn first-common-ancestor
```

Comment:
If you move a function into a util namespace, you should write a test for it. Otherwise it's too easy to drop test coverage. Also, I think `least-common-ancestor` is the common name for this, see https://en.wikipedia.org/wiki/Lowest_common_ancestor

Reply:
Missing test coverage - good point, will add. 

In general the tests feel pretty thin, but to some extent I'm waiting on seeing whether it gets used again before putting further effort into testing etc.

The rename came from finding "lowest" quite confusing in the context of value types, as I intuitively found myself thinking of the leaves as the "highest" (most specific) designations. Since this is now abstracted away from types the revert makes sense. TIL some standardized terminology.

Reply:
On reflection, I just realized that this function is not actually calculating a LCA - it returns the first member in the toposort walk, which I'm not sure will always be the same as the first member in a BFS.

Given this uncertainty, I'd rather keep the current name which is semantically more precise.

Reply:
Good point, the lowest common ancestor is not necessarily unique for a DAG. The name is good